plugins/node/instrumentation-amqplib/src/amqplib.ts

```diff
@@ -0,0 +1,733 @@
+  ) {
+    const normalizedExchange = normalizeExchange(exchange);
+
+    const span = self.tracer.startSpan(
+      `${normalizedExchange} -> ${routingKey} send`,
+      {
+        kind: SpanKind.PRODUCER,
+        attributes: {
+          ...channel.connection[CONNECTION_ATTRIBUTES],
+          [SemanticAttributes.MESSAGING_DESTINATION]: exchange,
+          [SemanticAttributes.MESSAGING_DESTINATION_KIND]:
+            MessagingDestinationKindValues.TOPIC,
+          [SemanticAttributes.MESSAGING_RABBITMQ_ROUTING_KEY]: routingKey,
+          [SemanticAttributes.MESSAGING_MESSAGE_ID]: options?.messageId,
+          [SemanticAttributes.MESSAGING_CONVERSATION_ID]:
+            options?.correlationId,
+        },
+      }
+    );
+    const modifiedOptions = options ?? {};
+    modifiedOptions.headers = modifiedOptions.headers ?? {};
+
+    propagation.inject(
+      trace.setSpan(context.active(), span),
+      modifiedOptions.headers
+    );
+
+    return { span, modifiedOptions };
+  }
+
+  private endConsumerSpan(
+    message: InstrumentationMessage,
+    isRejected: boolean | null,
+    operation: EndOperation,
+    requeue: boolean | undefined
+  ) {
+    const storedSpan: Span | undefined = message[MESSAGE_STORED_SPAN];
+    if (!storedSpan) return;
+    if (isRejected !== false) {
+      storedSpan.setStatus({
+        code: SpanStatusCode.ERROR,
+        message:
+          operation !== EndOperation.ChannelClosed &&
+          operation !== EndOperation.ChannelError
+            ? `${operation} called on message${
+                requeue === true
+                  ? ' with requeue'
+                  : requeue === false
+                  ? ' without requeue'
+                  : ''
+              }`
+            : operation,
+      });
+    }
+    this.callConsumeEndHook(storedSpan, message, isRejected, operation);
+    storedSpan.end();
+    message[MESSAGE_STORED_SPAN] = undefined;
+  }
+
+  private endAllSpansOnChannel(
+    channel: InstrumentationConsumeChannel,
+    isRejected: boolean,
+    operation: EndOperation,
+    requeue: boolean | undefined
+  ) {
+    const spansNotEnded: { msg: amqp.Message }[] =
+      channel[CHANNEL_SPANS_NOT_ENDED] ?? [];
+    spansNotEnded.forEach(msgDetails => {
+      this.endConsumerSpan(msgDetails.msg, isRejected, operation, requeue);
+    });
+    channel[CHANNEL_SPANS_NOT_ENDED] = [];
+  }
+
+  private callConsumeEndHook(
+    span: Span,
+    msg: amqp.ConsumeMessage,
+    rejected: boolean | null,
+    endOperation: EndOperation
+  ) {
+    if (!this._config.consumeEndHook) return;
+
+    safeExecuteInTheMiddle(
+      () => this._config.consumeEndHook!(span, { msg, rejected, endOperation }),
+      e => {
+        if (e) {
+          diag.error('amqplib instrumentation: consumerEndHook error', e);
+        }
+      },
+      true
+    );
+  }
+
+  private checkConsumeTimeoutOnChannel(channel: InstrumentationConsumeChannel) {
+    const currentTime = new Date().getTime();
```

Comment:
isn't `Date` sensitive regarding change of local system time? Maybe we should use the Otel build in performance timers instead which are guaranteed to be monotonic?

Reply:
fixed 👍